<a href="https://colab.research.google.com/github/akshaya-02ly/akshaya2260-nlp/blob/main/lab_11_2(nlp).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Train and evaluate a Logistic Regression model for text classification, including text preprocessing, TF-IDF feature extraction, and model performance evaluation. Please provide the text dataset for this task.

## Import Libraries

### Subtask:
Import all necessary libraries for data handling (pandas), text processing (re, nltk), machine learning model building (sklearn.linear_model.LogisticRegression, sklearn.feature_extraction.text.TfidfVectorizer, sklearn.model_selection.train_test_split), and model evaluation (sklearn.metrics.accuracy_score, precision_score, recall_score, f1_score, confusion_matrix). Ensure NLTK stopwords are downloaded.


**Reasoning**:
Import all the required libraries for data handling, text processing, machine learning model building, and model evaluation, and download the NLTK stopwords.



In [ ]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Download NLTK stopwords
nltk.download('stopwords')

print("Libraries imported and NLTK stopwords downloaded successfully.")

Libraries imported and NLTK stopwords downloaded successfully.


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Download NLTK stopwords
nltk.download('stopwords')

print("Libraries imported and NLTK stopwords downloaded successfully.")

Libraries imported and NLTK stopwords downloaded successfully.


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Download NLTK stopwords
nltk.download('stopwords')

print("Libraries imported and NLTK stopwords downloaded successfully.")

Libraries imported and NLTK stopwords downloaded successfully.


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Check if stopwords are already downloaded to prevent the 'already up-to-date' message in stderr
try:
    nltk.data.find('corpora/stopwords')
    print("NLTK stopwords are already downloaded.")
except LookupError:
    nltk.download('stopwords')
    print("NLTK stopwords downloaded for the first time.")

print("Libraries imported and NLTK stopwords ensured.")

NLTK stopwords are already downloaded.
Libraries imported and NLTK stopwords ensured.


## Load Dataset

### Subtask:
Load the text dataset into a pandas DataFrame.


**Reasoning**:
A sample text dataset will be created and loaded into a pandas DataFrame, as no dataset was provided. This DataFrame will have 'text' and 'label' columns with diverse entries and distinct categories, and then the first 5 rows will be displayed to verify the data.



In [ ]:
data = {
    'text': [
        'I love this product! It is amazing and works perfectly.',
        'This is a terrible experience. I regret buying it.',
        'The service was excellent, very helpful and fast.',
        'Not satisfied with the quality, it broke after a week.',
        'Highly recommend! Best purchase ever.',
        'Could be better, some features are missing.',
        'Fantastic movie, a must-watch for everyone.',
        'Absolutely disgusted by the poor customer support.',
        'Pretty good for the price, no major complaints.',
        'Worst decision ever to buy this, complete waste of money.'
    ],
    'label': [
        'positive', 'negative', 'positive', 'negative', 'positive',
        'negative', 'positive', 'negative', 'positive', 'negative'
    ]
}
df = pd.DataFrame(data)

print("Sample DataFrame created successfully.")
print(df.head())

Sample DataFrame created successfully.
                                                text     label
0  I love this product! It is amazing and works p...  positive
1  This is a terrible experience. I regret buying...  negative
2  The service was excellent, very helpful and fast.  positive
3  Not satisfied with the quality, it broke after...  negative
4              Highly recommend! Best purchase ever.  positive


## Preprocess Text Data

### Subtask:
Clean the text data by converting it to lowercase, removing punctuation, numbers, and special characters, and then removing common English stopwords using NLTK. Display a sample of the processed text.


**Reasoning**:
Define a function to preprocess text by converting to lowercase, removing unwanted characters, tokenizing, and removing stopwords. Apply this function to the 'text' column of the DataFrame and display the original and processed text.



In [ ]:
def preprocess_text(text):
    # Convert to lowercase
    text = text.lower()
    # Remove punctuation, numbers, and special characters, keeping only alphabetic characters and spaces
    text = re.sub(r'[^a-z\s]', '', text)
    # Tokenize the text
    words = text.split()
    # Get English stopwords
    stop_words = set(stopwords.words('english'))
    # Remove stopwords
    words = [word for word in words if word not in stop_words]
    # Join the processed words back into a string
    text = ' '.join(words)
    return text

# Apply the preprocessing function to the 'text' column
df['processed_text'] = df['text'].apply(preprocess_text)

print("Text preprocessing complete. Displaying a sample of original and processed text:")
print(df[['text', 'processed_text']].head())

Text preprocessing complete. Displaying a sample of original and processed text:
                                                text  \
0  I love this product! It is amazing and works p...   
1  This is a terrible experience. I regret buying...   
2  The service was excellent, very helpful and fast.   
3  Not satisfied with the quality, it broke after...   
4              Highly recommend! Best purchase ever.   

                         processed_text  
0  love product amazing works perfectly  
1     terrible experience regret buying  
2        service excellent helpful fast  
3          satisfied quality broke week  
4   highly recommend best purchase ever  


## Feature Extraction with TF-IDF

### Subtask:
Apply TF-IDF Vectorization to convert the preprocessed text data into numerical features. Display the size of the generated vocabulary.


**Reasoning**:
Initialize a TfidfVectorizer, fit it to the processed text data, transform the data into TF-IDF features, and then display the shape of the resulting matrix and the size of the generated vocabulary as per the instructions.



In [ ]:
vectorizer = TfidfVectorizer()
X_tfidf = vectorizer.fit_transform(df['processed_text'])

print("Shape of TF-IDF matrix (documents, vocabulary size):")
print(X_tfidf.shape)

vocabulary_size = len(vectorizer.vocabulary_)
print(f"Size of the generated vocabulary: {vocabulary_size}")

Shape of TF-IDF matrix (documents, vocabulary size):
(10, 46)
Size of the generated vocabulary: 46


## Explain TF-IDF Weighting

### Subtask:
Explain the concept and intuition behind TF-IDF (Term Frequency-Inverse Document Frequency) weighting.


## Explain TF-IDF Weighting

### Subtask:
Explain the concept and intuition behind TF-IDF (Term Frequency-Inverse Document Frequency) weighting.

#### Explanation of TF-IDF Weighting

**TF-IDF** (Term Frequency-Inverse Document Frequency) is a numerical statistic that reflects how important a word is to a document in a collection or corpus. It is widely used as a weighting factor in information retrieval and text mining. The intuition behind TF-IDF is that words that appear frequently in a document but rarely in the entire corpus are likely to be very relevant to that specific document.

TF-IDF is composed of two main components:

1.  **Term Frequency (TF)**
    *   **Concept**: Term Frequency measures how frequently a term (word) appears in a document. The more often a word appears, the more relevant it is considered to be to that document.
    *   **Intuition**: If a word occurs many times in a document, it probably describes the content of that document well. For example, in a document about 'apple', the word 'apple' would appear many times.
    *   **Calculation**: There are several ways to calculate TF, but a common method is the raw count of a term in a document, or the frequency normalized by the total number of words in the document to prevent bias towards longer documents. For example, `TF(t, d) = (Number of times term t appears in document d) / (Total number of terms in document d)`.

2.  **Inverse Document Frequency (IDF)**
    *   **Concept**: Inverse Document Frequency measures how important a term is across the entire corpus. It down-weights terms that appear very frequently across many documents and up-weights terms that appear rarely across the corpus.
    *   **Intuition**: Common words like "the", "is", "a" (stopwords) appear in almost every document and thus carry little unique information about a document's specific topic. Conversely, a word that appears in only a few documents is likely to be very distinctive and important for those documents. IDF helps to filter out these common, less informative words.
    *   **Calculation**: `IDF(t, D) = log_e(Total number of documents D / Number of documents containing term t)`. A small constant (usually 1) is often added to the denominator to prevent division by zero if a term does not appear in any document in the corpus.

### How TF-IDF Combines These Metrics

TF-IDF is calculated by multiplying TF and IDF:

`TF-IDF(t, d, D) = TF(t, d) * IDF(t, D)`

*   `t`: The term (word)
*   `d`: The specific document
*   `D`: The corpus (collection of documents)

**The intuition behind multiplying TF by IDF is to identify relevant and unique terms:**

*   If a word appears frequently in a *specific document* (high TF) but rarely across the *entire corpus* (high IDF), its TF-IDF score will be high. This indicates that the word is highly characteristic of that document and helps differentiate it from other documents.
*   If a word appears frequently in a *specific document* (high TF) but also frequently across the *entire corpus* (low IDF), its TF-IDF score will be moderate or low. This signifies that the word is common and not very distinctive for that document.
*   If a word appears rarely in a *specific document* (low TF), its TF-IDF score will be low, regardless of its IDF. This means the word is not central to that document's content.


## Split Dataset

### Subtask:
Split the preprocessed data into training and testing sets using `train_test_split` to prepare for model training and evaluation.


In [ ]:
X = X_tfidf
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print("Shapes of the training and testing sets:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Shapes of the training and testing sets:
X_train shape: (8, 46)
X_test shape: (2, 46)
y_train shape: (8,)
y_test shape: (2,)


## Train Logistic Regression Model

### Subtask:
Initialize and train a Logistic Regression model on the training data. Tune the `max_iter` parameter to ensure convergence.


**Reasoning**:
Initialize and train a Logistic Regression model using the training data, ensuring convergence with a suitable `max_iter` value and setting `random_state` for reproducibility.



In [ ]:
model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(X_train, y_train)

print("Logistic Regression model trained successfully.")

Logistic Regression model trained successfully.


## Explain Sigmoid Function Intuition

### Subtask:
Explain the intuition and role of the sigmoid activation function in Logistic Regression.


## Explain Sigmoid Function Intuition

### Subtask:
Explain the intuition and role of the sigmoid activation function in Logistic Regression.

#### Explanation of Sigmoid Function Intuition and Role in Logistic Regression

The **Sigmoid function**, also known as the **logistic function**, is a crucial component in Logistic Regression. Its primary role is to map any real-valued number into a value between 0 and 1, making it ideal for converting a linear combination of inputs into a probability.

1.  **Intuition and Mathematical Formula**

    The intuition behind the sigmoid function is to provide a smooth, S-shaped curve that can represent a probability. It is especially useful when we want to model the probability of a binary outcome (e.g., 0 or 1, true or false, positive or negative).

    The mathematical formula for the sigmoid function, often denoted as $\sigma(z)$, is:

    $$\sigma(z) = \frac{1}{1 + e^{-z}}$$

    Where:
    *   $z$ is the input to the function, which can be any real number.
    *   $e$ is Euler's number (approximately 2.71828).

2.  **Output Range and Relationship to Probabilities**

    The output range of the sigmoid function is always between **0 and 1** (exclusive). That is, $0 < \sigma(z) < 1$.

    *   As $z$ approaches positive infinity ($z \to \infty$), $e^{-z}$ approaches 0, so $\sigma(z)$ approaches 1.
    *   As $z$ approaches negative infinity ($z \to -\infty$), $e^{-z}$ approaches positive infinity, so $\sigma(z)$ approaches 0.
    *   When $z = 0$, $e^{-0} = 1$, so $\sigma(0) = \frac{1}{1 + 1} = 0.5$.

    This output range is perfectly suited for representing probabilities. In Logistic Regression, the output of the sigmoid function is interpreted as the probability that a given input belongs to a certain class (e.g., the probability of being 'positive').

3.  **Role in Logistic Regression**

    In Logistic Regression, the model first calculates a linear combination of the input features and their corresponding weights. This linear combination can be represented as $z = w_0 + w_1x_1 + w_2x_2 + ... + w_nx_n$, or more compactly as $z = \mathbf{w}^T\mathbf{x}$, where $\mathbf{w}$ is the vector of weights and $\mathbf{x}$ is the vector of input features.

    This $z$ value can be any real number, positive or negative. However, to transform this linear score into a probability for binary classification, we pass it through the sigmoid function.

    The sigmoid function then converts this $z$ value into a probability $P(Y=1|X)$, which represents the likelihood of the instance belonging to the positive class (e.g., label = 'positive'):

    $$P(Y=1|X) = \sigma(\mathbf{w}^T\mathbf{x}) = \frac{1}{1 + e^{-(\mathbf{w}^T\mathbf{x})}}$$

    If this probability is greater than a certain threshold (commonly 0.5), the instance is classified as belonging to the positive class; otherwise, it's classified as belonging to the negative class. This transformation allows Logistic Regression to perform binary classification by modeling the probability of an event occurring.

## Model Evaluation

### Subtask:
Evaluate the trained Logistic Regression model's performance on the test set. Calculate and display the accuracy, precision, recall, and F1-score. Also, generate and display a confusion matrix.


**Reasoning**:
The subtask requires evaluating the model's performance on the test set by calculating various metrics and generating a confusion matrix. I will use the trained model to make predictions on X_test, then compute accuracy, precision, recall, F1-score, and the confusion matrix, finally printing all results with descriptive labels.



In [ ]:
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, pos_label='positive')
recall = recall_score(y_test, y_pred, pos_label='positive')
f1 = f1_score(y_test, y_pred, pos_label='positive')
cm = confusion_matrix(y_test, y_pred)

print("Model Evaluation on Test Set:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")
print("\nConfusion Matrix:")
print(cm)

Model Evaluation on Test Set:
Accuracy: 0.5000
Precision: 0.5000
Recall: 1.0000
F1-Score: 0.6667

Confusion Matrix:
[[0 1]
 [0 1]]


## Final Task

### Subtask:
Summarize the entire process of developing and evaluating the Logistic Regression model for text classification, highlighting key steps and final performance metrics.


## Summary:

### Q&A
The task implicitly asks for a summary of the development and evaluation process of the Logistic Regression model for text classification, including key steps and final performance metrics.

### Data Analysis Key Findings

*   **Libraries and Setup**: All necessary libraries for data handling, text processing (NLTK stopwords), model building, and evaluation were successfully imported.
*   **Dataset Loading**: Since no external dataset was provided, a sample text dataset consisting of 10 entries with 'text' and 'label' columns ('positive'/'negative') was created for demonstration.
*   **Text Preprocessing**:
    *   A custom `preprocess_text` function was applied to the raw text.
    *   This function converted text to lowercase, removed punctuation, numbers, and special characters, and filtered out common English stopwords.
    *   The processed text was stored in a new 'processed\_text' column.
    *   The importance of these preprocessing steps for standardization, noise reduction, and focusing on informative words was explained.
*   **Feature Extraction**:
    *   TF-IDF (Term Frequency-Inverse Document Frequency) vectorization was used to convert the preprocessed text into numerical features.
    *   The resulting TF-IDF matrix had a shape of (10, 46), indicating 10 documents and a vocabulary size of 46 unique words.
    *   A detailed explanation of TF-IDF, including Term Frequency (TF) and Inverse Document Frequency (IDF) components, was provided, highlighting its role in weighting word importance.
*   **Data Splitting**: The dataset was split into training and testing sets, with 8 samples for training and 2 samples for testing (20% test size), using stratified sampling to maintain class distribution.
*   **Model Training**: A Logistic Regression model was initialized with `max_iter=1000` and `random_state=42` and successfully trained on the 8 training samples.
*   **Sigmoid Function Explanation**: The intuition and role of the sigmoid activation function in Logistic Regression, mapping linear combinations to probabilities between 0 and 1, were explained, including its mathematical formula $\sigma(z) = \frac{1}{1 + e^{-z}}$.
*   **Model Evaluation**: The trained model was evaluated on the 2 test samples, yielding the following performance metrics:
    *   Accuracy: 0.5000
    *   Precision (for 'positive' class): 0.5000
    *   Recall (for 'positive' class): 1.0000
    *   F1-Score (for 'positive' class): 0.6667
    *   The Confusion Matrix was:
        ```
        [[0 1]
         [0 1]]
        ```
        This indicates that out of the two test samples, one actual negative instance was predicted as positive (False Positive), and one actual positive instance was predicted as positive (True Positive).

### Insights or Next Steps

*   **Small Dataset Limitation**: The current model performance metrics are highly skewed and not reliable due to the extremely small dataset size (only 10 samples, with 2 in the test set). This setup is suitable for demonstration but not for drawing meaningful conclusions about model efficacy.
*   **Enhance Data for Robust Evaluation**: To properly evaluate the model, the next crucial step is to acquire and use a significantly larger and more representative text dataset. This would allow for more stable train-test splits, cross-validation, and a more accurate assessment of the Logistic Regression model's true performance.
